transport/cluster.go

```diff
@@ -118,18 +199,65 @@ func (c *Cluster) refreshTopology() error {
 				n.pool = pool
 			}
 		}
-		m[n.addr] = n
+
+		peers[n.addr] = n
+		all = append(all, n)
+		if err := insertNodeTokens(n, r, ring); err != nil {
+			return err
+		}
+
+		if _, ok := dcRacks[n.datacenter]; !ok {
+			dcRacks[n.datacenter] = 1
+		} else {
+			rackID := n.datacenter + n.rack
+			if _, ok := seenRacks[rackID]; !ok {
+				seenRacks[rackID] = true
+				dcRacks[n.datacenter]++
+			}
+		}
 	}
 
 	for k, v := range old {
-		if _, ok := m[k]; v.pool != nil && !ok {
+		if _, ok := peers[k]; v.pool != nil && !ok {
 			v.pool.Close()
 		}
 	}
-	c.setPeers(m)
+
+	c.setTopology(&topology{
+		peers:     peers,
+		dcRacks:   dcRacks,
+		allNodes:  all,
+		ring:      ring,
+		keyspaces: keyspaces,
+	})
+	// Discards request that came to us while refreshing.
+	select {
+	case <-c.refreshChan:
+	default:
+	}
 	return nil
 }
```

Comment:
I'm not quite sure why we need this?

Refactor to method and put close to where the put is.
In general this should be done in a for loop.

Reply:
It's not really necessary.
On the other hand, it seems logical to discard refresh topology requests that were sent during refreshing, because they are outdated.